torch/csrc/inductor/static_cuda_launcher.cpp

```diff
@@ -0,0 +1,462 @@
+  }
+}
+
+/* Load the CUDA kernel into memory (called during torch.compile), and
+  return a pointer to it (along with nregs and nspills).
+  Called in python as:
+  (function, n_regs, n_spills) = load_kernel(cubin_path, func_name,
+  sharedMemBytes)
+*/
+static PyObject* load_kernel(PyObject* self, PyObject* args) {
+  const char* filePath = nullptr;
+  const char* funcName = nullptr;
+  int sharedMemBytes = 0;
+  int n_regs = 0;
+  int n_spills = 0;
+  if (!PyArg_ParseTuple(args, "ssi", &filePath, &funcName, &sharedMemBytes)) {
+    return nullptr;
+  }
+  CUfunction func = nullptr;
+  try {
+    func = loadKernel(filePath, funcName, sharedMemBytes);
+    // Taken from triton/nvidia/backend/driver.c
+    AT_CUDA_DRIVER_CHECK(
+        nvrtc().cuFuncGetAttribute(&n_regs, CU_FUNC_ATTRIBUTE_NUM_REGS, func));
+    AT_CUDA_DRIVER_CHECK(nvrtc().cuFuncGetAttribute(
+        &n_spills, CU_FUNC_ATTRIBUTE_LOCAL_SIZE_BYTES, func));
+    n_spills /= 4;
+  } catch (const std::runtime_error& e) {
+    PyErr_SetString(PyExc_RuntimeError, e.what());
+    return nullptr;
+  }
+  // Return a tuple of CUFunction, n_regs, n_spills
+  return Py_BuildValue(
+      "(Kii)", reinterpret_cast<uint64_t>(func), n_regs, n_spills);
+}
+
+template <size_t NUM_ARGS>
+static PyObject* launch_kernel_inner(
+    CUfunction func,
+    int gridX,
+    int gridY,
+    int gridZ,
+    int numWarps,
+    int sharedMemBytes,
+    const char* argTypes,
+    PyObject* varArgs,
+    cudaStream_t cudaStream) {
+  try {
+    // Launch the kernel
+    // Prepare the arguments for the kernel
+    // We allocate 8 bytes per argument on the stack. We then allocate 8 more
+    // bytes to point to each 8 byte slot in argStorage, and pass that array of
+    // pointers to launchKernel.
+    std::array<uint64_t, NUM_ARGS> argStorage = {};
+    std::array<void*, NUM_ARGS> kernelArgs = {};
+    parseKernelArgs(varArgs, argTypes, argStorage, kernelArgs);
+
+    launchKernel(
+        func,
+        gridX,
+        gridY,
+        gridZ,
+        numWarps,
+        sharedMemBytes,
+        kernelArgs,
+        cudaStream);
+  } catch (const std::runtime_error& e) {
+    PyErr_SetString(PyExc_RuntimeError, e.what());
+    return nullptr;
+  }
+  Py_RETURN_NONE;
+}
+#define LAUNCH_KERNEL(N)  \
+  launch_kernel_inner<N>( \
+      func,               \
+      gridX,              \
+      gridY,              \
+      gridZ,              \
+      numWarps,           \
+      sharedMemBytes,     \
+      argTypes,           \
+      varArgs,            \
+      cudaStream)
+
+/**
+ *  Main entrypoint function called at runtime; called like this in python land:
+    launcher(
+      function, # CUfunction returned by load_kernel()
+      grid_x,
+      grid_y,
+      grid_z,
+      num_warps,
+      shared,
+      arg_tys, # e.g. "bO" for (int8_t, uint64_t)
+      args, # tuple of arguments passed to the kernel
+      stream,
+  )
+ *
+ */
+static PyObject* launch_kernel(PyObject* self, PyObject* args) {
```

Comment:
nit: might be worth marking this as noexcept - It's possible the compiler will complain if we call something that can throw...